File: electron/main/cross-device.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from 'node:http'
import { randomBytes } from 'node:crypto'
import { hostname, networkInterfaces } from 'node:os'
import { classify, makePreview } from '@shared/classify'
import type {
  CrossDeviceSendResult,
  CrossDeviceStatus,
} from '@shared/types'

const MAX_TEXT_BYTES = 100_000
const MAX_IMAGE_BYTES = 20 * 1024 * 1024
const CONNECTED_WINDOW_MS = 5_000

interface SharedItem {
  revision: number
  kind: 'text' | 'image'
  text: string | null
  image: Buffer | null
  preview: string
  sentAt: number
  source: 'desktop' | 'phone'
}

interface CrossDeviceDeps {
  onPhoneText(text: string): void
  onStatusChanged(): void
}

export class CrossDeviceService {
  private server: Server | null = null
  private token: string | null = null
  private port: number | null = null
  private address: string | null = null
  private lastSeenAt: number | null = null
  private latest: SharedItem | null = null
  private revision = 0

  constructor(private readonly deps: CrossDeviceDeps) {}

  async start(): Promise<CrossDeviceStatus> {
    if (this.server) return this.status()

    this.token = randomBytes(24).toString('hex')
    this.address = preferredLanAddress()
    if (!this.address) {
      this.token = null
      throw new Error('没有检测到可用的局域网 IPv4 地址')
    }
    const server = createServer((req, res) => this.handle(req, res))
    this.server = server

    try {
      await new Promise<void>((resolve, reject) => {
        const onError = (error: Error): void => reject(error)
        server.once('error', onError)
        server.listen(0, '0.0.0.0', () => {
          server.off('error', onError)
          resolve()
        })
      })
      const info = server.address()
      if (!info || typeof info === 'string') throw new Error('无法取得跨设备服务端口')
      this.port = info.port
      this.deps.onStatusChanged()
      return this.status()
    } catch (error) {
      this.server = null
      this.token = null
      this.port = null
      this.address = null
      throw error
    }
  }

  async stop(): Promise<CrossDeviceStatus> {
    const server = this.server
    this.server = null
    this.token = null
    this.port = null
    this.address = null
    this.lastSeenAt = null
    this.latest = null
    this.revision = 0

    if (server) {
      await new Promise<void>((resolve) => server.close(() => resolve()))
    }
    this.deps.onStatusChanged()
    return this.status()
  }

  status(): CrossDeviceStatus {
    const running = Boolean(this.server?.listening && this.token && this.port)
    const url =
      running && this.address && this.port && this.token
        ? `http://${this.address}:${this.port}/pair/${this.token}`
        : null
    return {
      running,
      url,
      pairCode: running && this.token ? this.token.slice(-6).toUpperCase() : null,
      connected:
        running &&
        this.lastSeenAt !== null &&
        Date.now() - this.lastSeenAt < CONNECTED_WINDOW_MS,
      lastSeenAt: this.lastSeenAt,
      lastSentAt: this.latest?.sentAt ?? null,
      lastSentPreview: this.latest?.preview ?? null,
    }
  }

  publishText(text: string, source: 'desktop' | 'phone' = 'desktop'): CrossDeviceSendResult {
    if (!this.server?.listening) return { ok: false, reason: 'not-running' }
    if (isSensitiveSyncText(text)) return { ok: false, reason: 'sensitive' }
    if (Buffer.byteLength(text, 'utf8') > MAX_TEXT_BYTES) {
      return { ok: false, reason: 'too-large' }
    }
    if (this.latest?.kind === 'text' && this.latest.text === text && this.latest.source === source) {
      return { ok: true }
    }

    this.latest = {
      revision: ++this.revision,
      kind: 'text',
      text,
      image: null,
      preview: makePreview(text),
      sentAt: Date.now(),
      source,
    }
    this.deps.onStatusChanged()
    return { ok: true }
  }

  publishImage(
    png: Buffer,
    preview = '剪贴板图片',
    source: 'desktop' | 'phone' = 'desktop',
  ): CrossDeviceSendResult {
    if (!this.server?.listening) return { ok: false, reason: 'not-running' }
    if (png.byteLength === 0 || png.byteLength > MAX_IMAGE_BYTES) {
      return { ok: false, reason: 'too-large' }
    }
    if (
      this.latest?.kind === 'image' &&
      this.latest.image?.equals(png) &&
      this.latest.source === source
    ) {
      return { ok: true }
    }
    this.latest = {
      revision: ++this.revision,
      kind: 'image',
      text: null,
      image: Buffer.from(png),
      preview,
      sentAt: Date.now(),
      source,
    }
    this.deps.onStatusChanged()
    return { ok: true }
  }

  private handle(req: IncomingMessage, res: ServerResponse): void {
    applySecurityHeaders(res)
    const requestUrl = new URL(req.url ?? '/', 'http://localhost')
    const token = this.token
    if (!token || !constantTimeTextEqual(requestUrl.pathname.split('/').pop() ?? '', token)) {
      respondText(res, 404, '配对已失效，请回到电脑重新生成二维码。')
      return
    }

    if (req.method === 'GET' && requestUrl.pathname.startsWith('/pair/')) {
      respondHtml(res, mobilePage(hostname(), token.slice(-6).toUpperCase()))
      return
    }

    if (req.method === 'GET' && requestUrl.pathname.startsWith('/api/state/')) {
      const wasConnected = this.status().connected
      this.lastSeenAt = Date.now()
      if (!wasConnected) this.deps.onStatusChanged()
      respondJson(res, 200, {
        ok: true,
        pairCode: token.slice(-6).toUpperCase(),
        latest: this.latest
          ? {
              revision: this.latest.revision,
              kind: this.latest.kind,
              text: this.latest.text,
              preview: this.latest.preview,
              sentAt: this.latest.sentAt,
              source: this.latest.source,
              imageUrl:
                this.latest.kind === 'image'
                  ? `/api/image/${token}?revision=${this.latest.revision}`
                  : null,
            }
          : null,
      })
      return
    }

    if (req.method === 'GET' && requestUrl.pathname.startsWith('/api/image/')) {
      const requestedRevision = Number(requestUrl.searchParams.get('revision'))
      if (
        this.latest?.kind !== 'image' ||
        !this.latest.image ||
        requestedRevision !== this.latest.revision
      ) {
        respondText(res, 404, '图片已经失效')
        return
      }
      res.writeHead(200, {
        'Content-Type': 'image/png',
        'Content-Length': this.latest.image.byteLength,
        'Content-Disposition': 'inline; filename="witchcat-clipboard.png"',
      })
      res.end(this.latest.image)
      return
    }

    if (req.method === 'POST' && requestUrl.pathname.startsWith('/api/send/')) {
      void this.readPhoneText(req, res)
      return
    }

    respondText(res, 404, 'Not found')
  }

  private async readPhoneText(req: IncomingMessage, res: ServerResponse): Promise<void> {
    try {
      const body = await readBody(req, MAX_TEXT_BYTES + 1_024)
      const payload = JSON.parse(body) as { text?: unknown }
      const text = typeof payload.text === 'string' ? payload.text : ''
      if (!text.trim()) {
        respondJson(res, 400, { ok: false, reason: 'empty' })
        return
      }
      const result = this.publishText(text, 'phone')
      if (!result.ok) {
        respondJson(res, result.reason === 'sensitive' ? 403 : 413, result)
        return
      }
      this.lastSeenAt = Date.now()
      this.deps.onPhoneText(text)
      respondJson(res, 200, { ok: true })
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error)
      respondJson(res, message === 'payload-too-large' ? 413 : 400, {
        ok: false,
        reason: message === 'payload-too-large' ? 'too-large' : 'invalid',
      })
    }
  }
}

export function isSensitiveSyncText(text: string): boolean {
  if (classify(text) === 'key') return true
  if (/(?:password|passwd|pwd|密码)\s*[:=：]\s*\S{4,}/i.test(text)) return true
  return text.split(/\r?\n/).some((line) => classify(line) === 'key')
}

function preferredLanAddress(): string | null {
  const candidates: { address: string; score: number }[] = []
  for (const [name, addresses] of Object.entries(networkInterfaces())) {
    for (const info of addresses ?? []) {
      if (info.family !== 'IPv4' || info.internal || !isPrivateIpv4(info.address)) continue
      let score = 0
      if (/wi-?fi|wlan|ethernet|以太网/i.test(name)) score += 50
      if (/virtual|vmware|vethernet|wsl|hyper-v|tailscale|loopback/i.test(name)) score -= 100
      if (info.address.startsWith('192.168.')) score += 20
      else if (info.address.startsWith('10.')) score += 10
      else score += 5
      candidates.push({ address: info.address, score })
    }
  }
  return candidates.sort((a, b) => b.score - a.score)[0]?.address ?? null
}

function isPrivateIpv4(address: string): boolean {
  if (/^10\./.test(address) || /^192\.168\./.test(address)) return true
  const match = /^172\.(\d+)\./.exec(address)
  if (!match) return false
  const second = Number(match[1])
  return second >= 16 && second <= 31
}

function constantTimeTextEqual(left: string, right: string): boolean {
  if (left.length !== right.length) return false
  let mismatch = 0
  for (let i = 0; i < left.length; i++) mismatch |= left.charCodeAt(i) ^ right.charCodeAt(i)
  return mismatch === 0
}

async function readBody(req: IncomingMessage, maxBytes: number): Promise<string> {
  const chunks: Buffer[] = []
  let bytes = 0
  for await (const chunk of req) {
    const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)
    bytes += buffer.byteLength
    if (bytes > maxBytes) throw new Error('payload-too-large')
    chunks.push(buffer)
  }
  return Buffer.concat(chunks).toString('utf8')
}

function applySecurityHeaders(res: ServerResponse): void {
  res.setHeader('Cache-Control', 'no-store')
  res.setHeader('X-Content-Type-Options', 'nosniff')
  res.setHeader('Referrer-Policy', 'no-referrer')
  res.setHeader('X-Frame-Options', 'DENY')
  res.setHeader(
    'Content-Security-Policy',
    "default-src 'self'; style-src 'unsafe-inline'; script-src 'unsafe-inline'; connect-src 'self'; img-src 'self' data:; base-uri 'none'; frame-ancestors 'none'",
  )
}

function respondJson(res: ServerResponse, status: number, value: unknown): void {
  res.writeHead(status, { 'Content-Type': 'application/json; charset=utf-8' })
  res.end(JSON.stringify(value))
}

function respondText(res: ServerResponse, status: number, value: string): void {
  res.writeHead(status, { 'Content-Type': 'text/plain; charset=utf-8' })
  res.end(value)
}

function respondHtml(res: ServerResponse, value: string): void {
  res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' })
  res.end(value)
}

function mobilePage(computerName: string, pairCode: string): string {
  const safeComputer = escapeHtml(computerName)
  const safeCode = escapeHtml(pairCode)
  return `<!doctype html>
<html lang="zh-CN">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1,viewport-fit=cover">
  <meta name="theme-color" content="#203638">
  <title>Witch Clipboard · 跨设备剪贴板</title>
  <style>
    :root{color-scheme:light;font-family:-apple-system,BlinkMacSystemFont,"Segoe UI","PingFang SC",sans-serif}
    *{box-sizing:border-box}
    body{margin:0;min-height:100vh;overflow-x:hidden;background:linear-gradient(145deg,#dce9e7 0%,#edf1ef 44%,#d7e1e4 100%);color:#172326}
    body:before,body:after{content:"";position:fixed;border-radius:999px;filter:blur(4px);pointer-events:none}
    body:before{width:270px;height:270px;right:-90px;top:-70px;background:linear-gradient(145deg,#91c6c0aa,#e8f4f1aa)}
    body:after{width:230px;height:230px;left:-100px;bottom:6%;background:linear-gradient(145deg,#9db9c3aa,#e3ecee88)}
    main{position:relative;z-index:1;width:min(100%,460px);margin:auto;padding:30px 18px 42px}
    .brand{margin:6px 4px 22px}.eyebrow{font-size:10px;font-weight:750;letter-spacing:.18em;color:#52706f;margin-bottom:6px}
    h1{font-size:25px;line-height:1.15;letter-spacing:-.03em;margin:0}.sub{font-size:12px;color:#657577;margin-top:8px}
    .dot{display:inline-block;width:7px;height:7px;border-radius:50%;background:#2f9b83;box-shadow:0 0 0 4px #2f9b8318;margin-right:8px}
    .card{background:#ffffff70;border:1px solid #ffffffa8;border-radius:24px;padding:18px;margin:14px 0;box-shadow:0 20px 55px #31484918,inset 0 1px 0 #ffffffb8;backdrop-filter:blur(26px) saturate(135%);-webkit-backdrop-filter:blur(26px) saturate(135%)}
    .label{font-size:11px;font-weight:700;letter-spacing:.08em;color:#617173;margin-bottom:11px}.content{min-height:104px;white-space:pre-wrap;word-break:break-word;font-size:16px;line-height:1.6;color:#162326}.received-image{display:block;width:100%;max-height:300px;object-fit:contain;border-radius:16px;background:#ffffff55;box-shadow:inset 0 0 0 1px #ffffff88}
    textarea{width:100%;min-height:116px;resize:vertical;border:1px solid #ffffffb5;border-radius:16px;padding:14px;font:inherit;color:#162326;outline:none;background:#ffffff68;box-shadow:inset 0 1px 4px #3348490b}
    textarea::placeholder{color:#849193}textarea:focus{border-color:#76a5a0;box-shadow:0 0 0 3px #4a8b8418;background:#ffffff88}
    button{width:100%;height:47px;border:1px solid #ffffff35;border-radius:15px;background:#203638;color:#fff;font-size:14px;font-weight:700;margin-top:13px;box-shadow:0 10px 26px #20363826;transition:transform .15s,background .15s}
    button:active{transform:scale(.985);background:#16292b}.meta{font-size:11px;color:#728184;margin-top:10px}
    .notice{font-size:11px;line-height:1.65;color:#68787a;text-align:center;padding:6px 20px}.toast{position:fixed;z-index:3;left:50%;bottom:24px;transform:translateX(-50%) translateY(20px);opacity:0;background:#17292be8;color:#fff;padding:11px 17px;border:1px solid #ffffff30;box-shadow:0 12px 36px #23363835;backdrop-filter:blur(18px);border-radius:999px;font-size:13px;transition:.2s;pointer-events:none}.toast.on{opacity:1;transform:translateX(-50%)}
    @media(prefers-color-scheme:dark){body{background:linear-gradient(145deg,#142224,#1c292b 48%,#182729);color:#eef5f4}body:before{background:#366b6770}body:after{background:#3d596570}.card{background:#ffffff12;border-color:#ffffff20;box-shadow:0 22px 60px #0005,inset 0 1px 0 #ffffff16}.eyebrow{color:#8cb7b2}h1,.content{color:#edf5f4}.sub,.label,.meta,.notice{color:#9eafaf}textarea{background:#ffffff10;border-color:#ffffff1c;color:#eff6f5}textarea:focus{background:#ffffff17;border-color:#679d97}textarea::placeholder{color:#829190}button{background:#dbe9e7;color:#172729;border-color:#fff8}}
  </style>
</head>
<body>
  <main>
    <header class="brand"><div class="eyebrow">WITCHCAT CONNECT</div><h1>跨设备剪贴板</h1><div class="sub"><span class="dot"></span>已连接 ${safeComputer} · 配对码 ${safeCode}</div></header>
    <section class="card">
      <div class="label">来自电脑</div>
      <div id="received" class="content">等待电脑发送文字、链接或图片…</div>
      <button id="copy">复制到手机</button>
      <div id="receivedMeta" class="meta">保持页面打开即可自动接收</div>
    </section>
    <section class="card">
      <div class="label">发送到电脑</div>
      <textarea id="outgoing" placeholder="在这里长按粘贴或输入文字"></textarea>
      <button id="send">发送并写入电脑剪贴板</button>
    </section>
    <div class="notice">仅在当前 Wi‑Fi 内传输；Key、Token 和密码默认拒绝发送。关闭电脑端跨设备功能后，本地址立即失效。</div>
  </main>
  <div id="toast" class="toast"></div>
  <script>
    const token = location.pathname.split('/').pop();
    const received = document.getElementById('received');
    const receivedMeta = document.getElementById('receivedMeta');
    const outgoing = document.getElementById('outgoing');
    const copyButton = document.getElementById('copy');
    let latestText = '';
    let latestKind = 'text';
    let latestImageUrl = '';
    let revision = -1;
    let toastTimer;
    function toast(text){const el=document.getElementById('toast');el.textContent=text;el.classList.add('on');clearTimeout(toastTimer);toastTimer=setTimeout(()=>el.classList.remove('on'),1800)}
    async function poll(){
      try{
        const res=await fetch('/api/state/'+token,{cache:'no-store'});
        if(!res.ok)throw new Error();
        const data=await res.json();
        if(data.latest&&data.latest.revision!==revision){
          revision=data.latest.revision;latestKind=data.latest.kind||'text';latestText=data.latest.text||'';latestImageUrl=data.latest.imageUrl||'';
          if(latestKind==='image'){
            const img=document.createElement('img');img.src=latestImageUrl;img.alt='来自电脑的剪贴板图片';img.className='received-image';
            received.replaceChildren(img);copyButton.textContent='打开或保存图片';
          }else{
            received.textContent=latestText;copyButton.textContent='复制到手机';
          }
          receivedMeta.textContent=(data.latest.source==='phone'?'已发送到电脑':'电脑刚刚发送')+' · '+new Date(data.latest.sentAt).toLocaleTimeString();
        }
      }catch{receivedMeta.textContent='连接已断开，请回到电脑重新配对'}
      setTimeout(poll,1200);
    }
    document.getElementById('copy').onclick=async()=>{
      if(latestKind==='image'){
        if(!latestImageUrl){toast('图片已经失效');return}
        const link=document.createElement('a');link.href=latestImageUrl;link.download='witchcat-clipboard.png';link.target='_blank';document.body.appendChild(link);link.click();link.remove();
        toast('已打开图片，也可以长按图片保存');return
      }
      if(!latestText){toast('还没有收到内容');return}
      try{
        if(navigator.clipboard&&window.isSecureContext)await navigator.clipboard.writeText(latestText);
        else{const box=document.createElement('textarea');box.value=latestText;box.style.position='fixed';box.style.opacity='0';document.body.appendChild(box);box.select();document.execCommand('copy');box.remove()}
        toast('已复制到手机剪贴板');
      }catch{toast('请长按上方文字手动复制')}
    };
    document.getElementById('send').onclick=async()=>{
      const text=outgoing.value.trim();if(!text){toast('请先粘贴或输入内容');return}
      try{
        const res=await fetch('/api/send/'+token,{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({text})});
        const data=await res.json();
        if(res.status===403)toast('Key 或 Token 已被安全拦截');
        else if(!res.ok)toast('发送失败，请稍后重试');
        else{outgoing.value='';toast('已发送到电脑剪贴板')}
      }catch{toast('连接已断开')}
    };
    poll();
  </script>
</body>
</html>`
}

function escapeHtml(value: string): string {
  return value.replace(/[&<>"']/g, (char) => {
    const entities: Record<string, string> = {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      '"': '&quot;',
      "'": '&#39;',
    }
    return entities[char]
  })
}

File: src-tauri/src/cross_device.rs
use std::{
    io::{Read, Write},
    net::{TcpListener, TcpStream, UdpSocket},
    sync::{
        atomic::{AtomicU64, Ordering},
        mpsc::Sender,
        Arc, Mutex,
    },
    thread,
    time::{Duration, SystemTime, UNIX_EPOCH},
};

use rand::RngCore;
use serde::Serialize;

const MAX_TEXT_BYTES: usize = 100_000;
const MAX_IMAGE_BYTES: usize = 20 * 1024 * 1024;

#[derive(Clone)]
enum SharedItem {
    Text {
        text: String,
        preview: String,
        sent_at: i64,
        revision: u64,
    },
    Image {
        png: Vec<u8>,
        preview: String,
        sent_at: i64,
        revision: u64,
    },
}

#[derive(Default)]
struct Inner {
    address: Option<String>,
    port: Option<u16>,
    token: Option<String>,
    last_seen_at: Option<i64>,
    latest: Option<SharedItem>,
    revision: u64,
}

pub struct CrossDevice {
    inner: Arc<Mutex<Inner>>,
    generation: Arc<AtomicU64>,
    phone_text: Sender<String>,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Status {
    pub running: bool,
    pub url: Option<String>,
    pub pair_code: Option<String>,
    pub connected: bool,
    pub last_seen_at: Option<i64>,
    pub last_sent_at: Option<i64>,
    pub last_sent_preview: Option<String>,
}

#[derive(Serialize)]
pub struct SendResult {
    pub ok: bool,
    pub reason: Option<&'static str>,
}

impl CrossDevice {
    pub fn new(phone_text: Sender<String>) -> Self {
        Self {
            inner: Arc::new(Mutex::new(Inner::default())),
            generation: Arc::new(AtomicU64::new(0)),
            phone_text,
        }
    }
    pub fn start(&self) -> Result<Status, String> {
        if self
            .inner
            .lock()
            .map_err(|_| "cross-device lock poisoned")?
            .port
            .is_some()
        {
            return Ok(self.status());
        }
        let address =
            lan_address().ok_or_else(|| "没有检测到可用的局域网 IPv4 地址".to_string())?;
        let listener = TcpListener::bind("0.0.0.0:0").map_err(|e| e.to_string())?;
        listener.set_nonblocking(true).map_err(|e| e.to_string())?;
        let port = listener.local_addr().map_err(|e| e.to_string())?.port();
        let mut random = [0u8; 24];
        rand::rng().fill_bytes(&mut random);
        let token = random
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect::<String>();
        {
            let mut inner = self
                .inner
                .lock()
                .map_err(|_| "cross-device lock poisoned")?;
            inner.address = Some(address);
            inner.port = Some(port);
            inner.token = Some(token);
        }
        let generation = self.generation.fetch_add(1, Ordering::AcqRel) + 1;
        let current = self.generation.clone();
        let inner = self.inner.clone();
        let phone = self.phone_text.clone();
        thread::spawn(move || {
            while current.load(Ordering::Acquire) == generation {
                match listener.accept() {
                    Ok((stream, _)) => handle(stream, &inner, &phone),
                    Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => {
                        thread::sleep(Duration::from_millis(40))
                    }
                    Err(_) => break,
                }
            }
        });
        Ok(self.status())
    }
    pub fn stop(&self) -> Status {
        self.generation.fetch_add(1, Ordering::AcqRel);
        *self.inner.lock().expect("cross-device lock poisoned") = Inner::default();
        self.status()
    }
    pub fn status(&self) -> Status {
        let inner = self.inner.lock().expect("cross-device lock poisoned");
        let running = inner.port.is_some() && inner.token.is_some();
        let url = match (&inner.address, inner.port, &inner.token) {
            (Some(a), Some(p), Some(t)) => Some(format!("http://{a}:{p}/pair/{t}")),
            _ => None,
        };
        let pair_code = inner
            .token
            .as_ref()
            .map(|t| t[t.len() - 6..].to_uppercase());
        let (last_sent_at, last_sent_preview) = match &inner.latest {
            Some(SharedItem::Text {
                sent_at, preview, ..
            })
            | Some(SharedItem::Image {
                sent_at, preview, ..
            }) => (Some(*sent_at), Some(preview.clone())),
            None => (None, None),
        };
        Status {
            running,
            url,
            pair_code,
            connected: running && inner.last_seen_at.is_some_and(|v| now_ms() - v < 5000),
            last_seen_at: inner.last_seen_at,
            last_sent_at,
            last_sent_preview,
        }
    }
    pub fn publish_text(&self, text: String) -> SendResult {
        if self.status().running == false {
            return fail("not-running");
        }
        if text.len() > MAX_TEXT_BYTES {
            return fail("too-large");
        }
        if crate::classify::is_sensitive_sync_text(&text) {
            return fail("sensitive");
        }
        let mut inner = self.inner.lock().unwrap();
        inner.revision += 1;
        inner.latest = Some(SharedItem::Text {
            preview: preview(&text),
            text,
            sent_at: now_ms(),
            revision: inner.revision,
        });
        SendResult {
            ok: true,
            reason: None,
        }
    }
    pub fn publish_image(&self, png: Vec<u8>, preview: String) -> SendResult {
        if !self.status().running {
            return fail("not-running");
        }
        if png.is_empty() || png.len() > MAX_IMAGE_BYTES {
            return fail("too-large");
        }
        let mut inner = self.inner.lock().unwrap();
        inner.revision += 1;
        inner.latest = Some(SharedItem::Image {
            png,
            preview,
            sent_at: now_ms(),
            revision: inner.revision,
        });
        SendResult {
            ok: true,
            reason: None,
        }
    }
}

fn handle(mut stream: TcpStream, inner: &Arc<Mutex<Inner>>, phone: &Sender<String>) {
    let _ = stream.set_read_timeout(Some(Duration::from_secs(2)));
    let mut data = Vec::new();
    let mut chunk = [0u8; 4096];
    loop {
        match stream.read(&mut chunk) {
            Ok(0) => break,
            Ok(n) => {
                data.extend_from_slice(&chunk[..n]);
                if request_complete(&data) {
                    break;
                }
                if data.len() > MAX_TEXT_BYTES + 8192 {
                    return;
                }
            }
            Err(_) => return,
        }
    }
    let request = String::from_utf8_lossy(&data);
    let first = request.lines().next().unwrap_or_default();
    let mut parts = first.split_whitespace();
    let method = parts.next().unwrap_or("");
    let path = parts.next().unwrap_or("/");
    let clean = path.split('?').next().unwrap_or(path);
    let token = clean.rsplit('/').next().unwrap_or("");
    let expected = inner.lock().ok().and_then(|v| v.token.clone());
    if expected.as_deref() != Some(token) {
        respond(
            &mut stream,
            404,
            "text/plain; charset=utf-8",
            b"Pairing expired",
        );
        return;
    }
    if method == "GET" && clean.starts_with("/pair/") {
        respond(
            &mut stream,
            200,
            "text/html; charset=utf-8",
            mobile_page().as_bytes(),
        );
        return;
    }
    if method == "GET" && clean.starts_with("/api/state/") {
        let mut state = inner.lock().unwrap();
        state.last_seen_at = Some(now_ms());
        let latest = match &state.latest {
            Some(SharedItem::Text {
                text,
                preview,
                sent_at,
                revision,
            }) => {
                serde_json::json!({"revision":revision,"kind":"text","text":text,"preview":preview,"sentAt":sent_at})
            }
            Some(SharedItem::Image {
                preview,
                sent_at,
                revision,
                ..
            }) => {
                serde_json::json!({"revision":revision,"kind":"image","text":null,"preview":preview,"sentAt":sent_at,"imageUrl":format!("/api/image/{token}?revision={revision}")})
            }
            None => serde_json::Value::Null,
        };
        let body = serde_json::to_vec(&serde_json::json!({"ok":true,"latest":latest})).unwrap();
        respond(&mut stream, 200, "application/json", &body);
        return;
    }
    if method == "GET" && clean.starts_with("/api/image/") {
        let image = inner.lock().ok().and_then(|v| match &v.latest {
            Some(SharedItem::Image { png, .. }) => Some(png.clone()),
            _ => None,
        });
        if let Some(png) = image {
            respond(&mut stream, 200, "image/png", &png)
        } else {
            respond(&mut stream, 404, "text/plain", b"Not found")
        };
        return;
    }
    if method == "POST" && clean.starts_with("/api/send/") {
        let body = request.split_once("\r\n\r\n").map(|v| v.1).unwrap_or("");
        let text = serde_json::from_str::<serde_json::Value>(body)
            .ok()
            .and_then(|v| v.get("text").and_then(|v| v.as_str()).map(str::to_string))
            .unwrap_or_default();
        if text.trim().is_empty() || crate::classify::is_sensitive_sync_text(&text) {
            respond(&mut stream, 403, "application/json", b"{\"ok\":false}");
            return;
        }
        let _ = phone.send(text.clone());
        let mut state = inner.lock().unwrap();
        state.last_seen_at = Some(now_ms());
        state.revision += 1;
        state.latest = Some(SharedItem::Text {
            text: text.clone(),
            preview: preview(&text),
            sent_at: now_ms(),
            revision: state.revision,
        });
        respond(&mut stream, 200, "application/json", b"{\"ok\":true}");
        return;
    }
    respond(&mut stream, 404, "text/plain", b"Not found")
}

fn request_complete(data: &[u8]) -> bool {
    let Some(pos) = data.windows(4).position(|v| v == b"\r\n\r\n") else {
        return false;
    };
    let header = String::from_utf8_lossy(&data[..pos]);
    let length = header
        .lines()
        .find_map(|line| {
            line.to_ascii_lowercase()
                .strip_prefix("content-length:")
                .and_then(|v| v.trim().parse::<usize>().ok())
        })
        .unwrap_or(0);
    data.len() >= pos + 4 + length
}
fn respond(stream: &mut TcpStream, status: u16, content_type: &str, body: &[u8]) {
    let reason = if status == 200 { "OK" } else { "Not Found" };
    let header=format!("HTTP/1.1 {status} {reason}\r\nContent-Type: {content_type}\r\nContent-Length: {}\r\nCache-Control: no-store\r\nX-Content-Type-Options: nosniff\r\nConnection: close\r\n\r\n",body.len());
    let _ = stream.write_all(header.as_bytes());
    let _ = stream.write_all(body);
}
fn lan_address() -> Option<String> {
    let socket = UdpSocket::bind("0.0.0.0:0").ok()?;
    socket.connect("192.0.2.1:80").ok()?;
    Some(socket.local_addr().ok()?.ip().to_string())
}
fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}
fn preview(text: &str) -> String {
    crate::classify::make_preview(text, 160)
}
fn fail(reason: &'static str) -> SendResult {
    SendResult {
        ok: false,
        reason: Some(reason),
    }
}
fn mobile_page() -> String {
    r#"<!doctype html><meta charset=utf-8><meta name=viewport content="width=device-width"><title>Witch Clipboard</title><style>body{font:16px system-ui;max-width:680px;margin:40px auto;padding:20px;background:#18131f;color:#eee}section{background:#2a2234;padding:20px;border-radius:18px;margin:16px 0}textarea{box-sizing:border-box;width:100%;min-height:140px;padding:12px}button{padding:12px 20px;background:#8b5cf6;color:white;border:0;border-radius:10px}img{max-width:100%}</style><h1>Witch Clipboard</h1><section><h3>来自电脑</h3><div id=received>等待内容…</div></section><section><h3>发送到电脑</h3><textarea id=out></textarea><p><button onclick=send()>发送</button></p></section><script>const token=location.pathname.split('/').pop();let revision=0;async function poll(){try{const s=await(await fetch('/api/state/'+token,{cache:'no-store'})).json();if(s.latest&&s.latest.revision!==revision){revision=s.latest.revision;received.innerHTML=s.latest.kind==='image'?'<img src="'+s.latest.imageUrl+'">':'<pre></pre>';if(s.latest.kind==='text')received.querySelector('pre').textContent=s.latest.text}}catch{}setTimeout(poll,1000)}async function send(){await fetch('/api/send/'+token,{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({text:out.value})});out.value=''}poll()</script>"#.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn local_pairing_server_sends_and_receives_text() {
        let (sender, receiver) = std::sync::mpsc::channel();
        let service = CrossDevice::new(sender);
        let status = service.start().unwrap();
        let url = status.url.unwrap();
        let tail = url.split_once("/pair/").unwrap();
        let port = tail.0.rsplit(':').next().unwrap();
        let token = tail.1;
        let mut stream = TcpStream::connect(format!("127.0.0.1:{port}")).unwrap();
        let body = r#"{"text":"from phone"}"#;
        write!(
            stream,
            "POST /api/send/{token} HTTP/1.1\r\nHost: localhost\r\nContent-Length: {}\r\n\r\n{}",
            body.len(),
            body
        )
        .unwrap();
        let mut response = String::new();
        stream.read_to_string(&mut response).unwrap();
        assert!(response.contains("200 OK"));
        assert_eq!(
            receiver.recv_timeout(Duration::from_secs(1)).unwrap(),
            "from phone"
        );
        assert!(service.publish_text("from desktop".into()).ok);
        assert!(!service.stop().running);
    }
}
